File: python/snakegame.py
#write a program for snake game in python
import pygame 
import random

# Initialize Pygame
pygame.init()

# Define the screen size
SCREEN_WIDTH = 640
SCREEN_HEIGHT = 480

# Set up the screen
screen = pygame.display.set_mode((SCREEN_WIDTH, SCREEN_HEIGHT))

# Define colors
WHITE = (255, 255, 255)
BLACK = (0, 0, 0)
RED = (255, 0, 0)

# Define the Snake class
class Snake:
    def __init__(self):
        self.body = [(SCREEN_WIDTH//2, SCREEN_HEIGHT//2)]
        self.direction = random.choice(['UP', 'DOWN', 'LEFT', 'RIGHT'])
        self.score = 0
        
    def move(self):
        # Get the current head position
        x, y = self.body[0]
        
        # Move the head in the current direction
        if self.direction == 'UP':
            y -= 10
        elif self.direction == 'DOWN':
            y += 10
        elif self.direction == 'LEFT':
            x -= 10
        elif self.direction == 'RIGHT':
            x += 10
            
        # Add the new head position to the body
        self.body = [(x, y)] + self.body[:-1]
        
    def turn(self, direction):
        # Change the direction of the snake
        if self.direction == 'UP' and direction != 'DOWN':
            self.direction = direction
        elif self.direction == 'DOWN' and direction != 'UP':
            self.direction = direction
        elif self.direction == 'LEFT' and direction != 'RIGHT':
            self.direction = direction
        elif self.direction == 'RIGHT' and direction != 'LEFT':
            self.direction = direction
            
    def eat_food(self, food):
        # Check if the snake has eaten the food
        if self.body[0] == food.position:
            self.body.append(self.body[-1])
            self.score += 10
            return True
        else:
            return False
        
    def hit_wall(self):
        # Check if the snake has hit a wall
        x, y = self.body[0]
        if x < 0 or x > SCREEN_WIDTH or y < 0 or y > SCREEN_HEIGHT:
            return True
        else:
            return False
        
    def hit_self(self):
        # Check if the snake has hit itself
        if self.body[0] in self.body[1:]:
            return True
        else:
            return False
        
    def draw(self):
        # Draw the snake on the screen
        for x, y in self.body:
            pygame.draw.rect(screen, BLACK, (x, y, 10, 10))
            
        # Draw the score on the screen
        font = pygame.font.Font(None, 36)
        text = font.render(f"Score: {self.score}", True, WHITE)
        screen.blit(text, (10, 10))
        
# Define the Food class
class Food:
    def __init__(self):
        self.position = self.random_position()
        
    def random_position(self):
        # Generate a random position for the food
        x = random.randrange(0, SCREEN_WIDTH, 10)
        y = random.randrange(0, SCREEN_HEIGHT, 10)
        return (x, y)
        
    def draw(self):
        # Draw the food on the screen
        pygame.draw.rect(screen, RED, (self.position[0], self.position[1], 10, 10))
        
# Create the Snake and Food objects
snake = Snake()
food = Food()

# Set up the game

File: C PROGRAMS/snakegame.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <conio.h>
#include <windows.h>

// Define the screen size
#define SCREEN_WIDTH 80
#define SCREEN_HEIGHT 25

// Define colors
#define WHITE 15
#define BLACK 0
#define RED 12

// Define the Snake struct
typedef struct Snake {
    int x, y;
    struct Snake *next;
} Snake;

// Define the Food struct
typedef struct Food {
    int x, y;
} Food;

// Initialize the console screen
void initialize_screen() {
    HANDLE hConsole = GetStdHandle(STD_OUTPUT_HANDLE);
    COORD screen_size = {SCREEN_WIDTH, SCREEN_HEIGHT};
    SetConsoleScreenBufferSize(hConsole, screen_size);
    SMALL_RECT window_size = {0, 0, SCREEN_WIDTH - 1, SCREEN_HEIGHT - 1};
    SetConsoleWindowInfo(hConsole, TRUE, &window_size);
}

// Clear the console screen
void clear_screen() {
    system("cls");
}

// Draw a character on the console screen
void draw_character(int x, int y, char c, int color) {
    HANDLE hConsole = GetStdHandle(STD_OUTPUT_HANDLE);
    COORD position = {x, y};
    SetConsoleCursorPosition(hConsole, position);
    SetConsoleTextAttribute(hConsole, color);
    putchar(c);
}

// Initialize the Snake
Snake *initialize_snake() {
    Snake *head = malloc(sizeof(Snake));
    head->x = SCREEN_WIDTH / 2;
    head->y = SCREEN_HEIGHT / 2;
    head->next = NULL;
    return head;
}

// Add a new segment to the Snake
void add_segment(Snake *head) {
    Snake *new_segment = malloc(sizeof(Snake));
    new_segment->x = head->x;
    new_segment->y = head->y;
    new_segment->next = head->next;
    head->next = new_segment;
}

// Move the Snake
void move_snake(Snake *head, int dx, int dy) {
    // Move the head
    head->x += dx;
    head->y += dy;
    
    // Move the body
    Snake *current_segment = head->next;
    int prev_x = head->x - dx;
    int prev_y = head->y - dy;
    while (current_segment != NULL) {
        int temp_x = current_segment->x;
        int temp_y = current_segment->y;
        current_segment->x = prev_x;
        current_segment->y = prev_y;
        prev_x = temp_x;
        prev_y = temp_y;
        current_segment = current_segment->next;
    }
}

// Check if the Snake has collided with a wall
int hit_wall(Snake *head) {
    if (head->x <= 0 || head->x >= SCREEN_WIDTH - 1 || head->y <= 0 || head->y >= SCREEN_HEIGHT - 1) {
        return 1;
    } else {
        return 0;
    }
}

// Check if the Snake has collided with itself
int hit_self(Snake *head) {
    Snake *current_segment = head->next;
    while (current_segment != NULL) {
        if (head->x == current_segment->x && head->y == current_segment->y) {
            return 1;
        }
        current_segment = current_segment->next;
    }
    return 0;
}

// Initialize the Food
Food initialize_food() {
    Food food;
    food.x = rand() % (SCREEN_WIDTH - 2);
}



